Setup shaders that render as well

matthijsgroen opened the pull request:
This is still a work in progress!

It is not yet finished. I create this pull request for now to get some early feedback :smile: 

TODO:
- [x] Test this with a built version
- [x] Create setup for images
- [x] Create setup for color
- [x] override default shader for sprites as demo
- [x] Refactor
- [x] Document the whole thing

matthijsgroen commented:
It works really great so far.

I can now override the default shader used for each component, and thus override the shader used for sprites.

https://github.com/matthijsgroen/game-play/blob/master/app/scripts/components/shaders/SpriteWithEffects.coffee

overrides the sprite color with same desaturation effect as images. I can now have enemies fly to the background and foreground in a nice way

matthijsgroen commented:
It's a rough draft for #1001 

There is still a lot to refactor, which I will do later this week.

kevinsimper commented:
I have not played around with shaders, do you have a good link to read about it? :smile:

matthijsgroen commented:
@kevinsimper: I've been exploring the code here: https://github.com/greggman/webgl-fundamentals/blob/master/webgl/webgl-2d-image-processing.html

Which is part of this website which I've been reading: http://webglfundamentals.org/

matthijsgroen commented:
Before I start documenting code, I think it would be better if @starwed or @mucaho did some sort of code review first?

matthijsgroen commented:
Anybody out there?

starwed commented:
This looks basically good to me -- it seems quite reasonable to group the various pieces needed for using a shader together in a single object.

I also don't see any issues that this would cause with the (somewhat stalled) work I was doing on adding multiple layers.

matthijsgroen commented:
@starwed I added documentation. I think it is ready for merging! (Should the whole thing be squashed into a single commit now?)

matthijsgroen committed:
Allow override of default Crafty shaders

Squashed commit of the following:

commit ba6ec3520d106ebebd7154505bf061241474408e
Author: Matthijs Groen <matthijs.groen@gmail.com>
Date:   Sat Jan 30 21:56:59 2016 +0100

    Fix line termination

commit c9d4e19eb3556f8b8b0ab35f170a6e693bf30223
Author: Matthijs Groen <matthijs.groen@gmail.com>
Date:   Sat Jan 30 21:55:38 2016 +0100

    Add documentation for overriding default shaders

commit 15e733257e51e032b3aeb6371c4faee551381cb5
Author: Matthijs Groen <matthijs.groen@gmail.com>
Date:   Sat Jan 23 18:30:56 2016 +0100

    Refactor usage of default shaders for an entity

commit 5cf1c35cdd4c57e328d85aa407511c0f7f1b979b
Author: Matthijs Groen <matthijs.groen@gmail.com>
Date:   Sat Jan 23 10:13:59 2016 +0100

    Fix argument check

commit 21e1014b8ea2ad0a3c6d6d2b90c53fcee3282e1e
Author: Matthijs Groen <matthijs.groen@gmail.com>
Date:   Sat Jan 23 10:09:59 2016 +0100

    Change order of build process

commit f95eab0d1da0ff763f42fdd3b1c284a2a2e4778a
Author: Matthijs Groen <matthijs.groen@gmail.com>
Date:   Sat Jan 23 09:57:01 2016 +0100

    Refactory color

commit b4a4cfdc61ee58a3b7dc0dd13b6c08655479f868
Author: Matthijs Groen <matthijs.groen@gmail.com>
Date:   Mon Jan 18 21:51:14 2016 +0100

    Update establish shader methods

commit e6d935dada0b778bcfb6ff60d788dd385ffcce70
Author: Matthijs Groen <matthijs.groen@gmail.com>
Date:   Mon Jan 18 21:26:44 2016 +0100

    Setup shaders that render as well

starwed commented:
Looks good to me, I'll run a couple of local tests, and if it renders ok go ahead and merge.

matthijsgroen commented:
:+1:

starwed commented on src/graphics/color.js:
```diff
@@ -135,16 +135,25 @@ Crafty.extend({
+    ],
+    function(e, _entity) {
+        e.program.writeVector("aColor",
+            entity._red/255,
```
Maybe you mean `entity`, not `_entity` here?  (or vice versa)

(Unfortunately we don't have unit tests for WebGL right now.)

starwed commented:
Other than the entity/_entity typo in "Color" looks like this works ok.

We don't have unit test coverage for webgl, and somewhat surprisingly `jshint` didn't catch this one either.   

(That said, I've been using [Visual Studio Code](https://code.visualstudio.com/) lately, and it does highlight the use of `entity` as an unknown symbol, which is a nice catch.  For working with Crafty I generally prefer it to Sublime Text out of the box, though I wasn't using many ST extensions.)

matthijsgroen commented:
It s actually a convention from other programming languages to prefix
unused arguments with an underscore.
Op za 6 feb. 2016 om 21:06 schreef Tim Martin notifications@github.com

> Other than the entity/_entity typo in "Color" looks like this works ok.
> 
> We don't have unit test coverage for webgl, and somewhat surprisingly
> jshint didn't catch this one either. That said, I've been using Visual
> Studio Code https://code.visualstudio.com/ lately, and it does
> highlight the use of entity as an unknown symbol, which is a nice catch.
> For working with Crafty I generally prefer it to Sublime Text out of the
> box, though I wasn't using many ST extensions.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/craftyjs/Crafty/pull/1003#issuecomment-180848146.

starwed commented:
> It s actually a convention from other programming languages to prefix unused arguments with an underscore.

Hmm, I haven't run into that before -- you do attempt to use the argument without the underscore here, though.

matthijsgroen commented:
Your correct! I'll fix this later today

matthijsgroen commented:
Updated. I've merged this version into my game for testing, and everything is fine.

starwed commented:
Looks good, merging.

mucaho commented:
Good job :+1: 
Custom shader programs, as well as custom vertex attributes allow for good customization.
For lightning and shadows you need multiple render passes and the ability to render to texture though. Has anyone got some ideas how to achieve that smoothly with current implementation?

mucaho commented:
Another thing to explore in the future is how to define custom uniforms alongside the internally used ones (unless this is already possible?) - they can be inefficiently emulated by copy-pasting it into every entity's vertex attribute though for now.
